### DISCUSSION TITLE
API Specification 정의 (MVP)

### DISCUSSION THREAD

**Discussion by cutehammond772** [SWM-Thlee/linked-paper#7] [General]
## 논의 주제
- MVP 버전의 각 세부 기능에 필요한 API 명세를 정의합니다.
- Front-End와 Back-End 간에 상호작용할 API 명세를 정의합니다.
- 각 세부 기능에 대해, 사전에 논의된 고민 사항 등을 고려합니다.

## 연관된 이슈
- #4

## 규칙
- 모든 API 명세는 일반적인 REST API 기준을 따릅니다.
- HTTP Method는 **GET**와 **POST**만을 사용합니다.
- Body의 경우 JSON Format을 따릅니다.
- Query String은 Route에 같이 표기하되, 설명에 세부 설명을 덧붙입니다.
- 위의 규칙을 토대로 Request와 Response를 모두 표기합니다.
- 별도의 HTTP Header가 필요한 경우 명시합니다. (ex. Cache-Control, Authorization 등)

## 명세 Format
- Route
- Headers (optional)
- Request Body (POST only)
- Response Body
  - 200 OK
  - 4xx, 5xx Error (optional)
- Description
- Trivia (optional)

**Comment by cutehammond772** [comment]
## 고려 사항 (정헌)
- 필터 정보를 요청으로 보낼 때, 각 필터 요소의 표현 방식
- 정렬 방식에 따른 요청 명세
- 검색 결과의 페이지네이션 방식
- 검색 결과의 리스트 뷰 또는 그래프 뷰로 표현해야 할 때, 둘 다 충족시키거나 확장성을 고려하도록 API 명세를 설계
- 유사도 컷을 넘어서 검색 결과를 요청할 때의 표현 방식
- 기타 API 명세 (피드백 보내기 등)

**Reply by cutehammond772** [reply]
## API 명세 목록

- Card Section
- Feedback
- Search
- Correlations
- Paper Details

**Comment by cutehammond772** [comment]
## API 명세가 필요한(or 연관된) 요소
- 메인 화면
  - 정보 관련 Card Section (서비스 소개, 패치 노트, 팀 소개 등)
  - 검색 창
  - 필터링 옵션
  - 피드백 보내기
- 검색 결과 화면
  - 필터링 옵션
  - 정렬 요소
  - 검색 결과

**Comment by cutehammond772** [comment]
## API: Card Section (v0)

### Route
> GET /info

### Response Body
```json
{
  "info": [
    {
      "icon": "info | summary | tip | question | success | warning | failure | bug | example | quote",
      "title": "string",
      "description": "string"
    },
  ]
}
```

### Description
- 메인 화면에 표시될 서비스 알림판 정보를 불러옵니다.
- `icon`의 경우 각 타입마다 프론트 단에서 특정 아이콘과 매치시키는 형태로 표현됩니다.

### Trivia
- 초반에 보여줄 데이터로는 이 서비스의 소개, 패치 노트, 소마 팀 소개 등이 있습니다.
- 이 데이터를 서버에 저장해 놓아도 되지만, 굳이 이런 API 필요 없이 프론트 단에서 하드코딩한 후 배포할 때마다 바꾸는 방법도 있습니다.

**Comment by cutehammond772** [comment]
## API: Feedback (v0)

### Route
> POST /feedback

### Request Body
```json
{
  "type": "... | others"
  "detail": "string"
  "description": "string"
}
```

### Response Body
```json
{}
```

### Description
- 사용자가 서비스 내의 특정 요소에 대해 피드백을 보낼 때 사용됩니다.
- `type`을 통해 어떤 요소에 대한 피드백인지 지정할 수 있습니다.
- 정해진 요소 이외의 피드백인 경우, `type`을 **others**로 지정한 후 `detail`에 어떤 요소인지 상세히 기록하여야 합니다.
- `detail`은 위의 경우에만 **required**이며, 이외의 경우에는 **optional**입니다.
- `description`에는 해당 요소에 대한 피드백 내용을 적어야 합니다.
- Response Body는 별도의 내용은 존재하지 않으며, Http Response Code만을 이용합니다.

### Trivia
- 사용자의 핑거프린트까지 수집하는 기능도 있으면 좋을 것 같습니다.
- 이러한 API를 만드는 대신, 구글 폼 등의 다른 플랫폼을 사용해도 되는 사안입니다.

**Comment by cutehammond772** [comment]
## API: Search (v0)

### Route
> GET /search?sentence={1}&sorting={2}&filter={3}&size={4}&page={5}&similarity_limit={6}

### Response Body
```json
{
  "count": "number",
  "status": "ok | similarity_limit_exceeded | not_found",
  "data": [
    {
      "id": "string",
      "title": "string",
      "authors": ["string", "string", "string"],
      "fields": ["string", "string", "string"],
      "journal": "string",
      "modified_date": "date",
      "abstraction": "string",

      "publish_date": "date",
      "citiation_count": "number",
      "similarity": "float",

      "origin_link": "string",
      "pdf_link": "string"
    },
  ]
}
```

### Description (전반)
- 사용자(클라이언트)가 검색 요청을 보낼 때 사용됩니다.
- 백 단에서는 쿼리 요청에 따른 **검색 결과**만 반환하며 나머지 후처리는 프론트 단에서 담당합니다.
- `count`는 총 검색 결과의 수를 나타냅니다. 
  - 일반적으로 `size`와 동일하지만, 유사도 제한 또는 검색 결과 부족으로 인하여 그 미만의 개수가 나올 수 있습니다.
- `status`는 검색 결과를 반환한 **이후**에 특정 상태에 도달하였는지의 여부를 나타냅니다.
  - 검색 결과가 유사도 제한에 걸린 경우, **SIMILARITY_LIMIT_EXCEEDED**가 반환됩니다.
  - 더이상 검색 결과가 존재하지 않는 경우, **NOT_FOUND**가 반환됩니다.
  - 다음 검색 결과가 존재하고, 현재 검색 결과가 정상적으로 반환된 경우 **OK**가 반환됩니다.
- `data`는 검색 결과가 `sorting` 순서로 정렬되어 나타납니다.

### Description (검색 결과)
- `id`는 해당 검색 결과(=논문)의 고유 ID를 나타냅니다.
  - DB에서 취급하는 PK일 수도 있고, OAI에서 취급하는 ID일 수도 있습니다.
- `fields`는 해당 논문이 취급하는 모든 분류를 나타냅니다.
  - 대분류인지 소분류인지 확실치 않으나, 일단 소분류를 기준으로 표기합니다.
- `modified_date`는 이 논문의 최종 수정 날짜를 나타냅니다.
  - 정렬 대상에는 포함되어 있지 않으며, 이 필드의 포함 여부 또한 결정된 사항은 아닙니다.
- `similarity`는 이 검색 결과의 유사도를 나타냅니다.
  - 내부 수치가 정확하게 반환되지만, 프론트 단에서는 이를 그대로 표시하지 않고 간략하게 표시됩니다.
- `origin_link`, `pdf_link`는 이 논문의 외부 링크를 나타냅니다.
  - 각 링크는 **optional**이며, 존재하는 경우에만 반환됩니다.

### Description (쿼리)
- `sentence`는 검색 창에 입력한 문장을 나타냅니다. 이 문장 내의 의미를 기반으로 검색을 수행합니다.
- `sorting`은 이 검색 결과의 정렬 순서를 나타냅니다. 기본 값은 유사도 순서입니다.
  - **similarity**, **citiation**, **recency**의 세 가지 옵션이 존재합니다.
- `filter`는 검색 수행 시 반영할 필터링 정보를 나타냅니다. 클라이언트에서 사전에 설정할 수 있습니다.
- `size`, `page`는 해당 검색 결과가 어느 정도 위치에 있는지를 나타냅니다.
  - 이 값들은 Pagination에 사용됩니다. 예를 들어, `size=20&page=2`이면 이 쿼리는 21번째~40번째 순서의 검색 결과를 반환합니다.
- `similarity_limit`은 이 쿼리가 유사도 제한을 고려하는지의 여부를 나타냅니다.
  - **true**일 때 유사도 제한이 활성화되며, **false**이면 제한 없이 검색 결과를 반환합니다.

### Description (필터)
- 해당 논문의 분류(**Field**), 저널(**Journal**), 게시일(**Date**)를 필터링 기준으로 합니다.
- JSON Format으로 나타내며, **Base64URL**로 인코딩 후 Query String에 포함됩니다.
- 해당 포맷은 클라이언트 단에서 저장되는 형식과 동일합니다.
- 특정 필터링 요소 내에 존재하는 기준들은 **OR**로 취급하며, 필터링 요소 간에는 **AND**로 취급합니다.

#### Example (JSON)
```json
{
  "field": {
    "cs": ["all"],
    "math": ["math.LO", "math.PR"]
  },
  "journal": ["Applied Network Science", "Cybersecurity"],
  "date": [
    {
      "start_date": "2008-10-01",
      "end_date": "2018-10-01"
    },
    {
      "start_date": "2022-01-01",
     "end-date": "2022-12-31"
    }
  ]
}
```

#### Example (Base64URL)
```base64
ewogICJmaWVsZCI6IHsKICAgICJjcyI6IFsiYWxsIl0sCiAgICAibWF0aCI6IFsibWF0aC5MTyIsICJtYXRoLlBSIl0KICB9LAogICJqb3VybmFsIjogWyJBcHBsaWVkIE5ldHdvcmsgU2NpZW5jZSIsICJDeWJlcnNlY3VyaXR5Il0sCiAgImRhdGUiOiBbCiAgICB7CiAgICAgICJzdGFydF9kYXRlIjogIjIwMDgtMTAtMDEiLAogICAgICAiZW5kX2RhdGUiOiAiMjAxOC0xMC0wMSIKICAgIH0sCiAgICB7CiAgICAgICJzdGFydF9kYXRlIjogIjIwMjItMDEtMDEiLAogICAgICJlbmQtZGF0ZSI6ICIyMDIyLTEyLTMxIgogICAgfQogIF0KfQ
```

**Reply by ljy2855** [reply]
깔끔하게 설계해주셨군요! 감사합니다!!

간단한 피드백 몇개만 남겨보아요
- page param -> index로 변경: 페이지 별 얼마나 불러올지 유동으로 변경될 수 있게, page보다는 몇번째부터 불러올지 지정하는게 좋을 것 같아요!
- `sentence` 라는 명칭보다 `query`라는 명칭은 어떨까요?
- `SIMILARITY_LIMIT_EXCEEDED` 라는 명칭보다 `LAST_PAGE` 같이 response 상태를 의미하는 명칭으로 정의하는게 좋아보여요!
- date은 여러 구간이 아니라 하나의 구간이었으면 해요.
- field로 마찬가지로 대분류 + 소분류 합쳐서 리스트로 처리했으면 해요. ex) cs.ALL , math.LO
- 필터의 depth를 고려하셔서 base64 인코딩을 고려하신것 같은데, 위의 수정사항을 고려해주신다면 인코딩 없이 진행해도 괜찮을 것 같아요.
```json
{
  "fields": ["cs.ALL", "math.LO"],
  "journals": ["Applied Network Science", "Cybersecurity"],
  "start_date" : "2022-01-01",
  "end_date" : "2023-01-01"
}
```

**Comment by cutehammond772** [comment]
## API: Search (v1)

### Route
> GET /search?{query}&{sorting}&{filter}&{size}&{index}&{similarity_limit}

### Response Body
```json
{
  "count": "number",
  "status": "OK | LAST_PAGE",
  "data": [
    {
      "id": "string",
      "title": "string",
      "authors": ["string", "string", "string"],
      "fields": ["string", "string", "string"],
      "journal": "string",
      "modified_date": "date",
      "abstraction": "string",

      "publish_date": "date",
      "reference_count": "number",
      "citiation_count": "number",
      "similarity": "float",

      "origin_link": "string",
      "pdf_link": "string"
    },
  ]
}
```

### Description (전반)
- 사용자(클라이언트)가 검색 요청을 보낼 때 사용됩니다.
- 백 단에서는 쿼리 요청에 따른 **검색 결과**만 반환하며 나머지 후처리는 프론트 단에서 담당합니다.
- `count`는 총 검색 결과의 수를 나타냅니다. 
  - 일반적으로 `size`와 동일하지만, 유사도 제한 또는 검색 결과 부족으로 인하여 그 미만의 개수가 나올 수 있습니다.
- `status`는 검색 결과를 반환한 **이후**에 특정 상태에 도달하였는지의 여부를 나타냅니다.
  - 더이상 검색 결과가 존재하지 않는 경우, **LAST_PAGE**가 반환됩니다. (유사도 제한에 걸린 경우도 마찬가지로 표시됩니다.)
  - 다음 검색 결과가 존재하고, 현재 검색 결과가 정상적으로 반환된 경우 **OK**가 반환됩니다.
- `data`는 검색 결과가 `sorting` 순서로 정렬되어 나타납니다.

### Description (검색 결과)
- `id`는 해당 검색 결과(=논문)의 고유 ID를 나타냅니다.
  - DB에서 취급하는 PK일 수도 있고, OAI에서 취급하는 ID일 수도 있습니다.
- `fields`는 해당 논문이 취급하는 모든 분류를 나타냅니다. 소분류를 기준으로 표기합니다.
- `modified_date`는 이 논문의 최종 수정 날짜를 나타냅니다.
- `similarity`는 이 검색 결과의 유사도를 나타냅니다.
  - 내부 수치가 정확하게 반환되지만, 웹 페이지에는 그대로 표시되지 않고 간략하게 표시됩니다.
- `origin_link`, `pdf_link`는 이 논문의 외부 링크를 나타냅니다. 각 링크는 **optional**이며, 존재하는 경우에만 명시됩니다.

### Description (쿼리)
- `query`는 검색 창 입력을 나타냅니다.
- `sorting`은 이 검색 결과의 정렬 순서를 나타냅니다. 기본 값은 유사도 순서입니다.
  - **none**, **similarity**, **citiation**, **recency**의 네 가지 옵션이 존재합니다.
- `filter`는 검색 수행 시 반영할 필터링 정보를 나타냅니다. 클라이언트에서 사전에 설정할 수 있습니다.
- `index`는 해당 검색 결과가 어느 정도 위치에 있는지를 나타냅니다.
- `size`를 통해 응답에 몇 개의 결과를 보내줄 것인지 요청 시 명시합니다.
- `similarity_limit`은 이 쿼리가 유사도 제한을 고려하는지의 여부를 나타냅니다.
  - **true**일 때 유사도 제한이 활성화되며, **false**이면 제한 없이 검색 결과를 반환합니다.

### Description (필터)
- 해당 논문의 분류(**Field**), 저널(**Journal**), 게시일(**Date**)를 필터링 기준으로 합니다.
- 쿼리스트링에서는 각 필드에 접두사로 `filter_`가 붙습니다.
- 특정 필터링 요소 내에 존재하는 기준들은 **OR**로 취급하며, 필터링 요소 간에는 **AND**로 취급합니다.

#### Example (JSON)
```json
{
  "field": ["cs.ALL", "math.LO"],
  "journal": ["Applied Network Science", "Cybersecurity"],
  "start_date": "2008-10-01",
  "end_date": "2018-10-01"
}
```

#### Example (Query String)
```
filter_field=cs.ALL,math.LO&filter_journal=Applied%20Network%20Science,Cybersecurity&filter_start_date=2008-10-01&filter_end_date=2018-10-01
```

**Comment by cutehammond772** [comment]
## API: Correlations (v0)

### Route

> GET /relation/{id}/{type}

### Response Body

```json
{
  "data": [
    {
      "id": "string",
      "title": "string",
      "authors": ["string", "string", "string"],
      "fields": ["string", "string", "string"],
      "journal": "string",
      "modified_date": "date",
      "abstraction": "string",

      "publish_date": "date",
      "reference_count": "number",
      "citiation_count": "number",
      "similarity": "float",

      "origin_link": "string",
      "pdf_link": "string"
    },
  ]
}
```

### Description (Route Segment)

 - `id`는 특정 논문의 ID를 나타냅니다.
 - `type`은 특정 논문과 어떻게 연관되어 있는지를 나타냅니다.
   - **reference**, **citiation**, **similar**의 세 종류가 존재합니다.

### Trivia

- 각 type마다 최대 몇 개씩 (또는 총 노드 갯수로 따진다면, 각 비중을 어떻게 할 것인지) 가져올 것인지 논의할 필요가 있습니다.
-

**Comment by cutehammond772** [comment]
## API: Paper Details (v0)

### Route

> GET /paper/{id}

### Response Body

```json
{
  "id": "string",
  "title": "string",
  "authors": ["string", "string", "string"],
  "fields": ["string", "string", "string"],
  "journal": "string",
  "modified_date": "date",
  "abstraction": "string",

  "publish_date": "date",
  "reference_count": "number",
  "citiation_count": "number",
  "similarity": "float",

  "origin_link": "string",
  "pdf_link": "string"
}
```

### Trivia

- 단독으로 Paper Details를 불러오는 상황은 어떤 경우가 있을까?
- 다른 API에서는 그냥 ID만 넣어 페이로드를 줄인 다음, 병렬적으로 이 API를 통해 로드하는 방법도 있지 않을까?
